File: Napakalaki_Ruby/player.rb
#encoding: utf-8

require_relative 'card_dealer.rb'
require_relative 'bad_consequence.rb'
require_relative 'treasure.rb'
require_relative 'treasure_kind.rb'
require_relative 'dice.rb'

module NapakalakiGame
  class Player
    @@MAXLEVEL = 10

    attr_reader :name
    attr_accessor :dead, :level, :visibleTreasures, :hiddenTreasures, :pendingBadConsequence

    def initialize (n)
      @name = n
      @dead = true
      @level = 1
      @visibleTreasures = Array.new
      @hiddenTreasures = Array.new
      @pendingBadConsequence = NumericBadConsequence.new("", 0, 0, 0)
    end

    def getName
      @name
    end

    def getVisibleTreasures
      @visibleTreasures
    end

    def getHiddenTreasures
      @hiddenTreasures
    end

    def createCopy
      nPlayer = Player.new(@name)
      nPlayer.dead = @dead
      nPlayer.level = @level
      nPlayer.visibleTreasures = Array.new(@visibleTreasures)
      nPlayer.hiddenTreasures = Array.new(@hiddenTreasures)
      nPlayer.pendingBadConsequence = @pendingBadConsequence
      nPlayer
    end

    def self.MAXLEVEL
      @@MAXLEVEL
    end

    def bringToLife
      @dead = false
    end

    def getCombatLevel
      suma = @level
      @visibleTreasures.each do |element|
        suma += element.bonus
      end
      suma
    end

    def getOponentLevel (m)
      m.combatLevel
    end

    def shouldConvert
      Dice.instance.nextNumber == 6
    end

    def incrementLevels (i)
      @level += i
    end

    def decrementLevels (i)
      if (@level - i) <= 1
        @level = 1
      else
        @level -= i
      end
    end

    def setPendingBadConsequence (b)
      @pendingBadConsequence = b
    end

    def applyPrize (m)
      nLevels = m.getLevelsGained
      incrementLevels(nLevels)
      nTreasures = m.getTreasuresGained

      if nTreasures > 0
        dealer = CardDealer.instance

        for i in 1..nTreasures do
          treasure = dealer.nextTreasure
          @hiddenTreasures << treasure
        end
      end
    end

    def applyBadConsequence (m)
      badConsequence = m.badConsequence
      nLevels = badConsequence.level
      decrementLevels(nLevels)
      pendingBad = badConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
      setPendingBadConsequence(pendingBad)
    end

    def canMakeTreasureVisible (t)
      if t.type == TreasureKind::ONEHAND
        howManyVisibleTreasures(TreasureKind::ONEHAND) < 2 && howManyVisibleTreasures(TreasureKind::BOTHHANDS) == 0
      elsif t.type == TreasureKind::BOTHHANDS
        howManyVisibleTreasures(TreasureKind::ONEHAND) == 0 && howManyVisibleTreasures(TreasureKind::BOTHHANDS) == 0
      else
        howManyVisibleTreasures(t.type) == 0
      end
    end

    def howManyVisibleTreasures (tKind)
      suma = 0
      @visibleTreasures.each do |element|
        if element.type == tKind
          suma += 1
        end
      end
      suma
    end

    def dieIfNoTreasures
      if @visibleTreasures.empty? && @hiddenTreasures.empty?
        @dead = true
      end
    end

    def combat (m)
      myLevel = getCombatLevel
      monsterLevel = getOponentLevel(m)

      if myLevel > monsterLevel
        applyPrize(m)
        if @level >= @@MAXLEVEL
          combatResult = CombatResult::WINGAME
        else
          combatResult = CombatResult::WIN
        end
      else
        applyBadConsequence(m)
        if shouldConvert
          combatResult = CombatResult::LOSEANDCONVERT
        else
          combatResult = CombatResult::LOSE
        end
      end
    end

    def makeTreasureVisible (t)
      canI = canMakeTreasureVisible(t)

      if canI
        @visibleTreasures << t
        @hiddenTreasures.delete(t)
      end
    end

    def discardVisibleTreasure (t)
      @visibleTreasures.delete(t)
      CardDealer.instance.giveTreasureBack(t)

      if @pendingBadConsequence!=nil && !@pendingBadConsequence.isEmpty
        @pendingBadConsequence.substractVisibleTreasure(t)
      end

      dieIfNoTreasures
    end

    def discardHiddenTreasure (t)
      @hiddenTreasures.delete(t)
      CardDealer.instance.giveTreasureBack(t)

      if @pendingBadConsequence!=nil && !@pendingBadConsequence.isEmpty
        @pendingBadConsequence.substractHiddenTreasure(t)
      end

      dieIfNoTreasures
    end

    def validState
      @pendingBadConsequence.isEmpty && (@hiddenTreasures.length <= 4)
    end

    def initTreasures
      dealer = CardDealer.instance
      dice = Dice.instance
      bringToLife
      treasure = dealer.nextTreasure
      @hiddenTreasures << treasure
      number = dice.nextNumber

      if number > 1
        treasure = dealer.nextTreasure
        @hiddenTreasures << treasure
      end

      if number == 6
        treasure = dealer.nextTreasure
        @hiddenTreasures << treasure
      end
    end

    def discardAllTreasures
      auxVisible = Array.new(@visibleTreasures)
      auxHidden = Array.new(@hiddenTreasures)

      auxVisible.each do |treasure|
        discardVisibleTreasure(treasure)
      end

      auxHidden.each do |treasure|
        discardHiddenTreasure(treasure)
      end
    end

    def to_s
      "#{@name}, nivel #{@level}"
    end

    protected :getCombatLevel, :getOponentLevel, :shouldConvert
    private :bringToLife, :incrementLevels, :decrementLevels, :setPendingBadConsequence, :applyPrize, :applyBadConsequence, :canMakeTreasureVisible, :howManyVisibleTreasures, :dieIfNoTreasures
  end
end

File: Napakalaki_Java/src/NapakalakiGame/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author francisco
 */
public class Player {
    static final int MAXLEVEL = 10;
    private String name;
    private int level;
    private boolean dead = true;
    private BadConsequence pendingBadConsequence;
    
    private ArrayList<Treasure> visibleTreasures = new ArrayList();
    private ArrayList<Treasure> hiddenTreasures = new ArrayList();
    
    public Player(String n){
        name = n;
        level = 1;
        pendingBadConsequence = new NumericBadConsequence("", 0, 0, 0);
    }
    
    public Player(Player pl){
        name = pl.getName();
        level = pl.getLevels();
        dead = pl.isDead();
        pendingBadConsequence = pl.getPendingBadConsequence();
        visibleTreasures = new ArrayList(pl.getVisibleTreasures());
        hiddenTreasures = new ArrayList(pl.getHiddenTreasures());
    }
    
    public String getName(){
        return name;
    }
    
    public int getLevels(){
        return level;
    }
    
    public ArrayList<Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }
    
    public ArrayList<Treasure> getHiddenTreasures(){
        return hiddenTreasures;
    }
    
    public BadConsequence getPendingBadConsequence(){
        return pendingBadConsequence;
    }
    
    public boolean isDead(){
        return dead;
    }
    
    private void bringToLife(){
        dead = false;
    }
    
    protected int getCombatLevel(){
        int suma = level;
        for (Treasure element : visibleTreasures){
            suma += element.getBonus();
        }
        return suma;
    }
    
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    protected boolean shouldConvert(){
        return Dice.getInstance().nextNumber() == 6;
    }
    
    private void incrementLevels(int i){
        level += i;
    }
    
    private void decrementLevels(int i){
        if ((level - i) <= 1){
            level = 1;
        }
        else{
            level -= i;
        }
    }
    
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence = b;
    }
    
    private void applyPrize(Monster m){
        int nLevels = m.getLevelsGained();
        incrementLevels(nLevels);
        int nTreasures = m.getTreasuresGained();

        if (nTreasures > 0){
            CardDealer dealer = CardDealer.getInstance();
            Treasure treasure;

            for (int i=0; i<nTreasures; i++){
                treasure = dealer.nextTreasure();
                hiddenTreasures.add(treasure);
            }
        }
    }
    
    private void applyBadConsequence(Monster m){
        BadConsequence badConsequence = m.getBadConsequence();
        int nLevels = badConsequence.getLevels();
        decrementLevels(nLevels);
        BadConsequence pendingBad = badConsequence.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
        setPendingBadConsequence(pendingBad);
    }
    
    private boolean canMakeTreasureVisible(Treasure t){
        if (t.getType() == TreasureKind.ONEHAND){
            return (howManyVisibleTreasures(TreasureKind.ONEHAND) < 2) && (howManyVisibleTreasures(TreasureKind.BOTHHANDS) == 0);
        }
        else if (t.getType() == TreasureKind.BOTHHANDS){
            return (howManyVisibleTreasures(TreasureKind.ONEHAND) == 0) && (howManyVisibleTreasures(TreasureKind.BOTHHANDS) == 0);
        }
        else{
            return howManyVisibleTreasures(t.getType()) == 0;
        }
    }
    
    private int howManyVisibleTreasures(TreasureKind tKind){
        int suma = 0;
    
        for (Treasure element : visibleTreasures){
            if (element.getType() == tKind){
                suma += 1;
            }
        }
        
        return suma;
    }
    
    private void dieIfNoTreasures(){
        if (visibleTreasures.isEmpty() && hiddenTreasures.isEmpty()){
            dead = true;
        }
    }
    
    public CombatResult combat(Monster m){
        int myLevel = getCombatLevel();
        int monsterLevel = getOponentLevel(m);
        CombatResult combatResult;

        if (myLevel > monsterLevel){
            applyPrize(m);
            
            if (level >= MAXLEVEL){
                combatResult = CombatResult.WINGAME;
            }
            else{
                combatResult = CombatResult.WIN;
            }
        }
        else{
            applyBadConsequence(m);
            
            if (shouldConvert()){
                combatResult = CombatResult.LOSEANDCONVERT;
            }
            else{
                combatResult = CombatResult.LOSE;
            }
        }
        
        return combatResult;
    }
    
    public void makeTreasureVisible(Treasure t){
        boolean canI = canMakeTreasureVisible(t);

        if (canI){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    
    public void discardVisibleTreasure(Treasure t){
        visibleTreasures.remove(t);
        CardDealer.getInstance().giveTreasureBack(t);

        if (pendingBadConsequence!=null && !pendingBadConsequence.isEmpty()){
            pendingBadConsequence.substractVisibleTreasure(t);
        }

        dieIfNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t){
        hiddenTreasures.remove(t);
        CardDealer.getInstance().giveTreasureBack(t);

        if (pendingBadConsequence!=null && !pendingBadConsequence.isEmpty()){
            pendingBadConsequence.substractHiddenTreasure(t);
        }

        dieIfNoTreasures();
    }
    
    public boolean validState(){
        System.out.println(pendingBadConsequence.toString());
        return pendingBadConsequence.isEmpty() && (hiddenTreasures.size() <= 4);
    }
    
    public void initTreasures(){
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        bringToLife();
        
        Treasure treasure = dealer.nextTreasure();
        hiddenTreasures.add(treasure);
        
        int number = dice.nextNumber();

        if (number > 1){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }

        if (number == 6){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
    }
    
    public void discardAllTreasures(){
        ArrayList<Treasure> auxVisible = new ArrayList(visibleTreasures);
        ArrayList<Treasure> auxHidden = new ArrayList(hiddenTreasures);

        for (Treasure treasure : auxVisible){
            discardVisibleTreasure(treasure);
        }

        for (Treasure treasure : auxHidden){
            discardHiddenTreasure(treasure);
        }
    }
    
    @Override
    public String toString(){
        return name + ", level " + Integer.toString(level);
    }
}
